datastream-bigquery/src/main/java/com/linkedin/datastream/bigquery/BigqueryBatchCommitter.java

```diff
@@ -86,47 +96,76 @@ public BigqueryBatchCommitter(VerifiableProperties properties) {
+    private synchronized void createOrUpdateTable(String destination) {
+        final String[] datasetTableNameRetention = destination.split("/");
 
-        this._destTableSchemas = new ConcurrentHashMap<>();
-        this._destTableCreated = new HashMap<>();
+        final Schema desiredTableSchema = _destTableSchemas.get(destination);
+
+        final long partitionRetentionDays = Long.parseLong(datasetTableNameRetention[2]);
+        final TimePartitioning timePartitioning;
+        if (partitionRetentionDays > 0) {
+            timePartitioning = TimePartitioning.of(TimePartitioning.Type.DAY, partitionRetentionDays * 86400000L);
+        } else {
+            timePartitioning = TimePartitioning.of(TimePartitioning.Type.DAY);
+        }
+
+        final TableId tableId = TableId.of(datasetTableNameRetention[0], sanitizeTableName(datasetTableNameRetention[1]));
+
+        final Optional<Table> optionalExistingTable = Optional.ofNullable(_bigquery.getTable(tableId));
+        if (optionalExistingTable.isPresent()) {
+            updateTable(destination, desiredTableSchema, timePartitioning, tableId, optionalExistingTable.get());
+        } else {
+            createTable(destination, desiredTableSchema, timePartitioning, tableId);
+        }
     }
 
-    private synchronized void createTableIfAbsent(String destination) {
-        if (_destTableCreated.containsKey(destination)) {
-            return;
+    private void updateTable(final String destination, final Schema desiredTableSchema,
+                             final TimePartitioning timePartitioning, final TableId tableId, final Table existingTable) {
+        final Schema existingTableSchema = Optional.ofNullable(existingTable.getDefinition().getSchema())
+                .orElseThrow(() -> new IllegalStateException(String.format("schema not defined for table: %s", tableId)));
+        if (!desiredTableSchema.equals(existingTableSchema)) {
+            final Schema evolvedSchema = _destTableSchemaEvolvers.get(destination).evolveSchema(existingTableSchema, desiredTableSchema);
+            if (!existingTableSchema.equals(evolvedSchema)) {
+                try {
+                    existingTable.toBuilder()
+                            .setDefinition(createTableDefinition(evolvedSchema, timePartitioning))
+                            .build().update();
+                } catch (BigQueryException e) {
+                    final Table currentTable = _bigquery.getTable(tableId);
+                    final Schema currentTableSchema = currentTable.getDefinition().getSchema();
+                    if (existingTableSchema.equals(currentTableSchema)) {
+                        LOG.error("Failed to update table {}", destination, e);
+                        throw e;
+                    } else {
+                        LOG.warn("Concurrent table schema update exception encountered for table {}. Retrying update with new base schema...", destination, e);
```

Comment:
Hi Chris, I am guessing concurrent update here is referring to the case where multi threads are updating schema at same time, is that right?

Reply:
Yes, the expectation here is that another thread updated the schema before we had a chance to apply our update, which we detect by comparing the previous table schema with the current schema. I actually should add an additional check here to see if the current schema matches the evolved schema, which is most likely to happen, in which case the table schema update is complete and successful. I'll work on adding that to the next commit.